crates/node/src/plugin_adapters/js_plugin_adapter/hooks/process_module.rs

```diff
@@ -0,0 +1,184 @@
+
+      Some(String::from_utf8_lossy(&code).to_string())
+    }
+    ModuleMetaData::Css(css_module_meta_data) => {
+      let (code, _) = codegen_css_stylesheet(&css_module_meta_data.ast, None, false);
+
+      Some(code)
+    }
+    ModuleMetaData::Html(html_module_meta_data) => {
+      Some(codegen_html_document(&html_module_meta_data.ast, false))
+    }
+    ModuleMetaData::Custom(_) => None,
+  })
+}
+
+fn convert_code_to_metadata(params: &mut PluginProcessModuleHookParam, code: String) -> Result<()> {
+  let filename = params.module_id.to_string();
+  match params.meta {
+    ModuleMetaData::Script(script_module_meta_data) => {
+      let ParseScriptModuleResult { ast, comments } = parse_module(
+        &filename,
+        &code,
+        // TODO: config should from config or process_module custom config
+        match params.module_type {
+          ModuleType::Js | ModuleType::Ts => Syntax::Es(Default::default()),
+          ModuleType::Jsx | ModuleType::Tsx => Syntax::Es(EsSyntax {
+            jsx: true,
+            ..Default::default()
+          }),
+          _ => Syntax::Es(Default::default()),
+        },
+        Default::default(),
+      )?;
+
+      script_module_meta_data.ast = ast;
+      script_module_meta_data.comments = comments.into()
+    }
+    ModuleMetaData::Css(css_module_meta_data) => {
+      let ParseCssModuleResult { ast, comments } = parse_css_stylesheet(&filename, Arc::new(code))?;
+
+      css_module_meta_data.ast = ast;
+      css_module_meta_data.comments = comments.into();
+    }
+    ModuleMetaData::Html(html_module_meta_data) => {
+      let v = parse_html_document(&filename, Arc::new(code))?;
+
+      html_module_meta_data.ast = v;
+    }
+    ModuleMetaData::Custom(_) => {
+      return Ok(());
+    }
+  }
+
+  Ok(())
+}
+
+impl JsPluginProcessModuleHook {
+  new_js_plugin_hook!(
+    PluginProcessModuleHookFilters,
+    JsPluginProcessModuleHookFilters,
+    CompatiblePluginProcessModuleHookParams,
+    PluginProcessModuleHookResult
+  );
+
+  pub fn call(
+    &self,
+    param: &mut PluginProcessModuleHookParam,
+    ctx: Arc<CompilationContext>,
+  ) -> Result<Option<()>> {
+    if self.filters.module_types.contains(param.module_type)
+      || self
+        .filters
+        .resolved_paths
+        .iter()
+        .any(|m| m.is_match(param.module_id.to_string().as_str()))
+    {
+      let Some(result) = format_module_metadata_to_code(param.meta)? else {
+        return Ok(None);
+      };
+
+      let Some(result) = self
+        .tsfn
+        .call::<CompatiblePluginProcessModuleHookParams, PluginProcessModuleHookResult>(
+          CompatiblePluginProcessModuleHookParams {
+            module_id: param.module_id.clone(),
+            module_type: param.module_type.clone(),
+            content: result,
+          },
+          ctx,
+          None,
+        )?
+      else {
+        return Ok(None);
+      };
+
+      convert_code_to_metadata(param, result.content)?;
```

Comment:
Can we only parse and codegen ast once? we can call all process_module hook of js plugins in serial , only codegen  the ast at start and parse content back when all plugins finished

Reply:
Requires a destructive update of the plugin driver, not changed in 1.x